### ISSUE TITLE
Unexpected exception when using include statement in handler

### ISSUE THREAD

**Issue by SteveTalbot** [ansible/ansible#18357]
##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME

- Handlers
- Include statement

##### ANSIBLE VERSION

Problem occurs with:
```
ansible 2.2.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```

Problem does not occur if I roll back to:
```
ansible 2.1.1.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```

##### CONFIGURATION

Using ansible.cfg as supplied by package maintainer of http://ppa.launchpad.net/ansible/ansible/ubuntu

##### OS / ENVIRONMENT

- Running from Ubuntu 16.04
- Used as a Vagrant provisioner
- Managing a VirtualBox VM running Ubuntu 16.04

##### SUMMARY

When Ansible runs a handler which is an include statement, the first included step fails with the error:
`ERROR! Unexpected Exception: 'NoneType' object has no attribute 'register'`

The ability to create a multi-step handler is really useful, so it's a shame this has stopped working for us.

##### STEPS TO REPRODUCE

Here is an example of a task that notifies the handler that causes the problem:
```
- name: Set CouchDB log directory and bind to correct IP address
  ini_file: dest=/etc/couchdb/local.ini
    section="{{ item.section }}"
    option="{{ item.option }}"
    value="{{ item.value }}"
  with_items:
    - { section: log, option: file, value: "{{ logPath }}/couchdb/couch.log" }
    - { section: httpd, option: bind_address, value: "{{ couchdbBindAddress }}" }
  notify: restart couchdb
  tags:
    - configure
```

The `restart couchdb` handler is defined in `handlers/main.yml` like this:
```
- name: restart couchdb
  include: "{{ role_path }}/tasks/restart.yml"
```

And the included file `tasks/restart.yml` contains a sequence of tasks like this:
```
# Detect whether CouchDB was already running
- name: Detect whether CouchDB is running
  uri: url="http://localhost:5984/_all_dbs" method=GET status_code=200
  register: couchdb_running_test
  ignore_errors: yes
  tags:
    - always

# Only restart if the service should be started on boot, or if it was already running
- name: Trigger CouchDB service restart
  service: name=couchdb state=restarted
  when: "couchdbStartOnBoot or couchdb_running_test|success"
  tags:
    - always
```

When the handler is executed, Ansible correctly includes file `tasks/restart.yml`, but fails to run the first task.  Full stack trace below.

We have discovered that the first task in the included file always fails in Ansible 2.2.0, regardless of what that task is.  So for example we can change the included file to start with a really simple task like this, and it still fails with the same error:
```
- name: Do nothing
  debug: msg="This should really work"
  tags:
    - always
```

It works as expected when we rolled back to Ansible 2.1.1.

##### EXPECTED RESULTS

The expected behaviour is:
- Handler is notified
- Handler runs at end of play
- "include" statement is processed; file is included
- All tasks in the included file are executed in sequence

Expected output (from Ansible 2.1.1) is:
```
RUNNING HANDLER [couchdb : restart couchdb] ************************************
included: /home/xxx/git/xxx/xxx/config/ansible/galaxy_roles/couchdb/tasks/restart.yml for default

RUNNING HANDLER [couchdb : uri] ************************************************
fatal: [default]: FAILED! => {"changed": false, "content": "", "failed": true, "msg": "Status code was not [200]: Request failed: <urlopen error [Errno 111] Connection refused>", "redirected": false, "status": -1, "url": "http://localhost:5984/_all_dbs"}
...ignoring

RUNNING HANDLER [couchdb : service] ********************************************
changed: [default]
```

##### ACTUAL RESULTS

Actual behaviour in Ansible 2.2.0 is:
- Handler is notified
- Handler runs at end of play
- "include" statement is processed; file is included
- First task in the included file fails with exception as follows

```
RUNNING HANDLER [couchdb : restart couchdb] ************************************
included: /home/xxx/git/xxx/xxx/config/ansible/galaxy_roles/couchdb/tasks/restart.yml for default

RUNNING HANDLER [couchdb : uri] ************************************************
Using module file /usr/lib/python2.7/dist-packages/ansible/modules/core/network/basics/uri.py
<127.0.0.1> ESTABLISH SSH CONNECTION FOR USER: vagrant
<127.0.0.1> SSH: EXEC ssh -vvv -o UserKnownHostsFile=/dev/null -o IdentitiesOnly=yes -o IdentitiesOnly=yes -o ControlMaster=auto -o ControlPersist=60s -o StrictHostKeyChecking=no -o Port=2222 -o 'IdentityFile="/home/xxx/git/xxx/xxx/.vagrant/machines/default/virtualbox/private_key"' -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=vagrant -o ConnectTimeout=30 -o ControlPath=/home/xxx/.ansible/cp/ansible-ssh-%h-%p-%r 127.0.0.1 '/bin/sh -c '"'"'( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1478255214.96-274210896519514 `" && echo ansible-tmp-1478255214.96-274210896519514="` echo $HOME/.ansible/tmp/ansible-tmp-1478255214.96-274210896519514 `" ) && sleep 0'"'"''
<127.0.0.1> PUT /tmp/tmp4o4etY TO /home/vagrant/.ansible/tmp/ansible-tmp-1478255214.96-274210896519514/uri.py
<127.0.0.1> SSH: EXEC scp -vvv -o UserKnownHostsFile=/dev/null -o IdentitiesOnly=yes -o IdentitiesOnly=yes -o ControlMaster=auto -o ControlPersist=60s -o StrictHostKeyChecking=no -o Port=2222 -o 'IdentityFile="/home/xxx/git/xxx/xxx/.vagrant/machines/default/virtualbox/private_key"' -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=vagrant -o ConnectTimeout=30 -o ControlPath=/home/xxx/.ansible/cp/ansible-ssh-%h-%p-%r /tmp/tmp4o4etY '[127.0.0.1]:/home/vagrant/.ansible/tmp/ansible-tmp-1478255214.96-274210896519514/uri.py'
<127.0.0.1> ESTABLISH SSH CONNECTION FOR USER: vagrant
<127.0.0.1> SSH: EXEC ssh -vvv -o UserKnownHostsFile=/dev/null -o IdentitiesOnly=yes -o IdentitiesOnly=yes -o ControlMaster=auto -o ControlPersist=60s -o StrictHostKeyChecking=no -o Port=2222 -o 'IdentityFile="/home/xxx/git/xxx/xxx/.vagrant/machines/default/virtualbox/private_key"' -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=vagrant -o ConnectTimeout=30 -o ControlPath=/home/xxx/.ansible/cp/ansible-ssh-%h-%p-%r 127.0.0.1 '/bin/sh -c '"'"'chmod u+x /home/vagrant/.ansible/tmp/ansible-tmp-1478255214.96-274210896519514/ /home/vagrant/.ansible/tmp/ansible-tmp-1478255214.96-274210896519514/uri.py && sleep 0'"'"''
<127.0.0.1> ESTABLISH SSH CONNECTION FOR USER: vagrant
<127.0.0.1> SSH: EXEC ssh -vvv -o UserKnownHostsFile=/dev/null -o IdentitiesOnly=yes -o IdentitiesOnly=yes -o ControlMaster=auto -o ControlPersist=60s -o StrictHostKeyChecking=no -o Port=2222 -o 'IdentityFile="/home/xxx/git/xxx/xxx/.vagrant/machines/default/virtualbox/private_key"' -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=vagrant -o ConnectTimeout=30 -o ControlPath=/home/xxx/.ansible/cp/ansible-ssh-%h-%p-%r -tt 127.0.0.1 '/bin/sh -c '"'"'/usr/bin/python /home/vagrant/.ansible/tmp/ansible-tmp-1478255214.96-274210896519514/uri.py; rm -rf "/home/vagrant/.ansible/tmp/ansible-tmp-1478255214.96-274210896519514/" > /dev/null 2>&1 && sleep 0'"'"''
ERROR! Unexpected Exception: 'NoneType' object has no attribute 'register'
the full traceback was:

Traceback (most recent call last):
  File "/usr/bin/ansible-playbook", line 103, in <module>
    exit_code = cli.run()
  File "/usr/lib/python2.7/dist-packages/ansible/cli/playbook.py", line 159, in run
    results = pbex.run()
  File "/usr/lib/python2.7/dist-packages/ansible/executor/playbook_executor.py", line 154, in run
    result = self._tqm.run(play=play)
  File "/usr/lib/python2.7/dist-packages/ansible/executor/task_queue_manager.py", line 282, in run
    play_return = strategy.run(iterator, play_context)
  File "/usr/lib/python2.7/dist-packages/ansible/plugins/strategy/linear.py", line 221, in run
    results.extend(self._execute_meta(task, play_context, iterator))
  File "/usr/lib/python2.7/dist-packages/ansible/plugins/strategy/__init__.py", line 838, in _execute_meta
    self.run_handlers(iterator, play_context)
  File "/usr/lib/python2.7/dist-packages/ansible/plugins/strategy/__init__.py", line 703, in run_handlers
    result = self._do_handler_run(handler, handler.get_name(), iterator=iterator, play_context=play_context)
  File "/usr/lib/python2.7/dist-packages/ansible/plugins/strategy/__init__.py", line 772, in _do_handler_run
    notified_hosts=included_file._hosts[:],
  File "/usr/lib/python2.7/dist-packages/ansible/plugins/strategy/__init__.py", line 743, in _do_handler_run
    host_results = self._wait_on_pending_results(iterator)
  File "/usr/lib/python2.7/dist-packages/ansible/plugins/strategy/__init__.py", line 534, in _wait_on_pending_results
    results = self._process_pending_results(iterator)
  File "/usr/lib/python2.7/dist-packages/ansible/plugins/strategy/__init__.py", line 316, in _process_pending_results
    if original_task.register:
AttributeError: 'NoneType' object has no attribute 'register'
Ansible failed to complete successfully. Any error output should be
visible above. Please fix these errors and try again.
```

Some directory and user names in the stack trace have been replaced with "xxx".

Rolling back to Ansible 2.1.1 is a workaround for now, but a solution for Ansible >=2.2 would be very welcome!

**Comment by nilsfr** [comment]
I can confirm that this happens if you include tasks in a handler.

```
- name: reload haproxy
  include: assemble_and_reload.yaml
```

**Comment by leseb** [comment]
Same error here, includes in handler break with ansible 2.2
This PR has the issue: https://github.com/ceph/ceph-ansible/pull/995/

**Comment by jimi-c** [comment]
Confirmed I'm seeing this when the include is dynamic in a handler, looking into it now.

**Comment by jimi-c** [comment]
Closing This Ticket
===================

Hi!

We believe the above commit should resolve this problem for you. This will also be included in the next release.

If you continue seeing any problems related to this issue, or if you have any further questions, please let us know by stopping by one of the two mailing lists, as appropriate:

   * https://groups.google.com/forum/#!forum/ansible-project - for user questions, tips, and tricks
   * https://groups.google.com/forum/#!forum/ansible-devel - for strategy, future planning, and questions about writing code

Because this project is very active, we're unlikely to see comments made on closed tickets, but the mailing list is a great way to ask questions, or post if you don't think this particular issue is resolved.

Thank you!

**Comment by leseb** [comment]
@jimi-c 
Do we know which Ansible version will get that fix? Any plan to backport it to 2.2 or 2.2.0.1 maybe? I see the commit was pushed on 'devel'.

Thanks!

**Comment by jimi-c** [comment]
@leseb it will be included in 2.2.1 (currently in RC1, this patch will be in RC2).

**Comment by leseb** [comment]
@jimi-c thanks!
